### PR TITLE
redesign, upgrade to Bootstrap 4 and Fork Awesome

### PR DESCRIPTION
**Pull request by thetric**
Upgrades to Bootstrap 4. BS4 [dropped Glyphicons](https://getbootstrap.com/docs/4.1/migration/#components) so I've replace them with [~Font Awesome~](https://fontawesome.com/icons) [Fork Awesome](https://github.com/ForkAwesome/Fork-Awesome) (which also contains more icons).

# TODOs:
- [x] apply styling to inputs
- [x] apply styling for Bootstrap's custom forms (radio and check boxes)
- [x] bug(options/nav): "general" stays active after tab switch
- [x] ~redesign "connected databases" with [cards](https://getbootstrap.com/docs/4.1/components/card/)?~ (later, future PR)
- [x] ~refactor(icons): replace font awesome web fonts + css with svg&js (smaller bundle?)~ (use Fo**rk**Awesome instead)
- [x] refactor(icons): replace Fo**nt**Awesome icons with Fo**rk**Awesome (see https://github.com/keepassxreboot/keepassxc-browser/pull/322#issuecomment-433696065)
- [x] use old green color theme (https://github.com/keepassxreboot/keepassxc-browser/pull/322#issuecomment-433695850)

# files:
- [x] `options/options.html`
- [x] `popups/popup.html`
- [x] `popups/popup_httpauth.html`
- [x] `popups/popup_multiple-fields.html`
- [x] `popups/popup_remember.html`
- [x] `options/shortcuts.html`

# bugs:
- [x] options/site-prefs: no warnings displayed if value is invalid
- [ ] options/site prefs: cannot change "Username-only Detection" on freshly added entry, although the changes are reflected to the model after a reload
- [ ] option/general: "Default group for saving new passwords:" displays "Default: null KeePassXC-Browser Passwords" with `null`?

# Previews:

image

image

image

image

image

### CODE REVIEW

**Comment by thetric** [comment]
WIP of the "General" tab

screenshot_2018-10-27_19-47-01

**Comment by phoerious** [comment]
I like it. I've always wanted to give it a slight makeover myself, but never had the time.

**Comment by droidmonkey** [comment]
Oh sexy!

File: keepassxc-browser/manifest.json [hunk 1]
```diff
@@ -100,11 +100,5 @@
         "http://*/*",
         "https://api.github.com/"
     ],
-    "applications": {
```
Comment (varjolintu) [inline keepassxc-browser/manifest.json]: Why did you remove this?

**Comment by varjolintu** [comment]
Looks very nice! Btw, are we going to keep the older green color sheme? The new navbar is bright green.

**Comment by varjolintu** [comment]
Some things to keep eye on:
https://github.com/twbs/bootstrap/pull/23586
https://forkawesome.github.io/Fork-Awesome/

**Comment by thetric** [comment]
This is mostly done (except the navbar bug).

Due to the change the primary color to the old green the 'success' color was redefined as light purple. Does this look okay? (E.g. in the popup after KeepassXC is connected)

**Comment by sebast889** [comment]
Shouldn't the auto-fill warning come under the "Automatically fill in single-credential entries"? In the above screenshot it looks like it is for the Automatically retrieve credentials" option.

Also, can the main header tabs like "General" be a darker green color when it is the active tab (just like it is for the current layout)?

**Comment by thetric** [comment]
> Shouldn't the auto-fill warning come under the "Automatically fill in single-credential entries"? In the above screenshot it looks like it is for the Automatically retrieve credentials" option.

Done:

screenshot_2018-12-22_18-31-55

> Also, can the main header tabs like "General" be a darker green color when it is the active tab (just like it is for the current layout)?

Or how do you like this: (The nav bar entries should have the same width)

image

**Comment by sebast889** [comment]
First one looks better to me, second has too much white. Is it not possible to match the current navbar layout?

**Comment by thetric** [comment]
Fixed the active nav item issue ("general" tab stayed active) and refactored the navbar to look like the current one:

screenshot_2019-01-26_15-32-18

**Comment by thetric** [comment]
Todo: remove script tag for redundant web extensions polyfills in Firefox-only page `shortcuts.html`

**Comment by droidmonkey** [comment]
Is this ready for review?

**Comment by thetric** [comment]
@droidmonkey yes

**Comment by varjolintu** [comment]
@thetric Please check my reviews.

Btw, I was also wondering if the increase of the font sizes is actually wanted here. The popup dialog seems quite huge with much of wasted space.

I also advice you to check the popup content using both Firefox and Chromium and see that it looks identical. Now there are some strange differences in the button sizes etc.

**Comment by thetric** [comment]
Reduced font size in popups and fixed inconsistent widths across browsers:

Firefox:

screenshot_2019-02-16_15-02-27

Chrome:

screenshot_2019-02-16_15-02-52


There is still is little difference in the top button bar but this shouldn't be too bad.

**Comment by sebast889** [comment]
Any reason for the color changes in buttons ("Choose custom login fields for this page" and "Redetect login fields")? I feel the current layouts colors are easier to identify (no 2 buttons have the same color and no transparent button that differs from the look of the other buttons).

**Comment by droidmonkey** [comment]
Yah I'm not a fan of the transparent button. Also the text should be shortened, something like "Select Login Fields" with a nice icon (mouse hovering over form?)

Move settings over to the far right, eliminate the text, the icon is appropriate. Add text to the lock icon "Lock Databases". There is a lot to be done to this dialog.

**Comment by thetric** [comment]
Firefox:

screenshot_2019-02-23_14-02-23

Chrome:

screenshot_2019-02-23_14-03-39

**Comment by phoerious** [comment]
The identifier should be on its own line. Breaking it into two lines looks rather ugly.

**Comment by droidmonkey** [comment]
I really like the new buttons!

**Comment by thetric** [comment]
The navbar in the options page looks weird on <991px:

screenshot_2019-02-23_14-22-55

**Comment by thetric** [comment]
> The identifier should be on its own line. Breaking it into two lines looks rather ugly.

@phoerious Done

screenshot_2019-02-23_14-29-27

**Comment by thetric** [comment]
Fixed the responsive behavior of the navbar:

screenshot_2019-02-23_15-05-53


screenshot_2019-02-23_15-04-38

File: keepassxc-browser/popups/popup.css [hunk 2]
```diff
@@ -1,51 +1,37 @@
 body {
-    font-family: sans-serif;
     overflow-x: hidden;
     background-color: #eee;
```
Comment (varjolintu) [inline keepassxc-browser/popups/popup.css]: Maybe the popup background color should be white, just like the Settings page.

File: keepassxc-browser/options/shortcuts.js [hunk 3]
```diff
@@ -65,53 +65,60 @@ async function updateKeys() {
...
         await browser.commands.update({
             name: shortcut,
-            shortcut: document.querySelector('#' + shortcut).value
+            shortcut: document.querySelector(`#${shortcut}`).value
         });
-        createBanner('success', shortcut);
-    } catch (e) {
-        console.log('Cannot change shortcut: ' + e);
-        createBanner('danger', shortcut);
+        createAlert('success', shortcut);
```
[the remark this answers is not in the document]
Comment (thetric) [reply keepassxc-browser/options/shortcuts.js]: The `alertArea` will be moved to the end of the page.
I'll investigate your bug.

File: keepassxc-browser/popups/popup.css (hunk 2 above)
Comment (thetric) [reply keepassxc-browser/popups/popup.css]: Yeah, agreed

File: build-bootstrap.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+const autoprefixer = require('autoprefixer');
+const postcssPresetEnv = require('postcss-preset-env');
+const { writeFile } = require('fs');
+const { render: renderScss } = require('node-sass');
+const postcss = require('postcss');
+
+const inFile = 'keepassxc-browser/styles/bootstrap.scss';
+const outFile = 'keepassxc-browser/styles/bootstrap.css';
+
+renderScss(
```
[the remark this answers is not in the document]
Comment (varjolintu) [reply build-bootstrap.js]: If we are going to use the build script, the CSS output file should be minimized. Also, I'm not sure if the extension is going to pass the AMO/CWB audit if there's .scss files inside the package.

File: keepassxc-browser/popups/popup.css (hunk 2 above)
Comment (varjolintu) [reply keepassxc-browser/popups/popup.css]: Nice. Everything else is already looking much better!

**Comment by thetric** [comment]
Restyled the popup when multiple logins are available:

Unbenannt

File: build-bootstrap.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (thetric) [reply build-bootstrap.js]: The AMO Wiki [doesn't recommend minifying code](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/Source_Code_Submission):

> Obfuscating or minifying your code isn't recommended. It rarely protects your code, as these techniques can be reversed using tools such as JSNice and JS Beautifier. And, unlike the advantage that minified code offers web pages loaded over the internet, as extension code is loaded from a local source the performance benefits are not significant.

[According to the AMO Wiki](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/Source_Code_Submission#Provide_your_extension_source_code) the `bootstrap.scss` file shouldn't present a problem:

> ## When must source code be uploaded?
> You must upload your extension’s source code when its code was created using:
> [...]
> * any other custom tool that takes files, applies pre-processing, and generates file(s) to include in the extension.

However it would be necessary to provide build instructions and [upload the sources when publishing a new addon version](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/Source_Code_Submission#How_to_upload_source_code):

> ## Provide build instructions
> To reproduce the build, the reviewer runs the instructions you provided and then uses a diff tool to compare the generated sources to those in the extension. There must be no differences. The easiest way to provide the build instructions is to include a README file with the submitted source code. [...]
>
> [...]
> When using npm, yarn, or other package management tools that support it, be sure to include the lockfile, for example, `package-lock.json`. Otherwise, reviewers may use a different version resulting in differences between the generated code and that in the extension.
> 
> [...] However, you don’t need to describe how to install common tools such as npm or node.

[the remark this answers is not in the document]
Comment (varjolintu) [reply build-bootstrap.js]: Still, the .scss file is not needed inside the extension.

**Comment by varjolintu** [comment]
FYI, we are going to merge #416 soon. All changes in that PR should be included in this one too. Most of the relevant changes are in the Settings page. And those should be quite easy to copy.

Keep up the good work!

**Comment by thetric** [comment]
What needs to be done before merging this? Except integrating the changes from #416

**Comment by varjolintu** [comment]
@thetric https://github.com/keepassxreboot/keepassxc-browser/pull/396 would be needed also.

**Comment by varjolintu** [comment]
FYI: https://blog.jquery.com/2019/04/10/jquery-3-4-0-released/

But maybe we shouldn't upgrade because: 

> Specifically, jQuery 3.4.0 is deprecating :first, :last, :eq, :even, :odd, :lt, :gt, and :nth. When we remove Sizzle, we’ll replace it with a small wrapper around querySelectorAll, and it would be almost impossible to reimplement these selectors without a larger selector engine.

**Comment by thetric** [comment]
How does this relate to this issue? It does not upgrade jQuery

**Comment by varjolintu** [comment]
@thetric Yes, it doesn't. It was just FYI, if you were thinking about it.

**Comment by thetric** [comment]
Merged the latest changes from `develop`. Ready for review

File: package.json [hunk 5]
```diff
@@ -20,5 +26,10 @@
   "bugs": {
     "url": "https://github.com/keepassxreboot/keepassxc-browser/issues"
   },
-  "homepage": "https://github.com/keepassxreboot/keepassxc-browser#readme"
+  "homepage": "https://github.com/keepassxreboot/keepassxc-browser#readme",
+  "browserslist": [
```
Comment (varjolintu) [inline package.json]: Does this follow the minimum versions of Firefox specified in manifest.json?

File: keepassxc-browser/options/options.html [hunk 6]
```diff
@@ -2,439 +2,1034 @@
...
-          <li class="active"><a href="#general-settings" tabindex="1"><span data-i18n="optionsMenuGeneral"></span></a></li>
-          <li><a href="#connected-databases" tabindex="2"><span data-i18n="optionsMenuConnectedDatabases"></span></a></li>
-          <li><a href="#custom-fields" tabindex="3"><span data-i18n="optionsMenuCustomFields"></span></a></li>
-          <li><a href="#site-preferences" tabindex="4"><span data-i18n="optionsMenuSitePreferences"></span></a></li>
-          <li><a href="#about" tabindex="5"><span data-i18n="optionsMenuAbout"></span></a></li>
-        </ul>
+      <h3><img src="/icons/keepassxc_48x48.png" alt="logo" /> KeePassXC-Browser</h3>
+
+      <nav class="navbar navbar-expand-lg navbar-dark bg-primary text-white mb-4">
+        <button
```
Comment (varjolintu) [inline keepassxc-browser/options/options.html]: Is there any reason for this file to be in different syntax than the rest of the HTML files? Why are the elements divided in multiple lines?

File: keepassxc-browser/options/options.html [hunk 7]
```diff
@@ -2,439 +2,1034 @@
...
-      <script src="../translate.js" defer></script>
+    <meta charset="UTF-8" />
+    <link rel="stylesheet" href="/styles/bootstrap.css" />
+    <link rel="stylesheet" href="options.css" />
+    <link rel="stylesheet" href="../fonts/fork-awesome.min.css" />
+    <script type="text/javascript" src="../browser-polyfill.min.js"></script>
+    <script type="text/javascript" src="../global.js"></script>
+    <script type="text/javascript" src="../jquery-3.3.1.min.js"></script>
+    <script type="text/javascript" src="bootstrap.min.js"></script>
+    <script type="text/javascript" src="options.js"></script>
```
Comment (varjolintu) [inline keepassxc-browser/options/options.html]: The `type` is not needed with HTML5. See the original file.

File: keepassxc-browser/options/options.html [hunk 8]
```diff
@@ -2,439 +2,1034 @@
...
+                  <span data-i18n="optionsDefault" i18n-placeholder="-1"></span>,
+                  <span data-i18n="optionsRecommended" i18n-placeholder="2000"></span>
+                </small>
+                <button
+                  class="btn btn-sm btn-primary mt-2"
+                  id="blinkMinTimeoutButton"
+                  type="button"
+                >
+                  <i class="fa fa-floppy-o" aria-hidden="true"></i>
+                  <span data-i18n="optionsButtonSave"></span>
```
Comment (varjolintu) [inline keepassxc-browser/options/options.html]: Warnings are not displayed when the value is invalid.

File: keepassxc-browser/options/options.html [hunk 9]
```diff
@@ -2,439 +2,1034 @@
...
+            <div class="input-group mb-3" id="manualUrlGroup">
+              <input
+                id="manualUrl"
+                type="url"
+                class="form-control"
+                pattern="ftp://.*|http://.*|https://.*"
+                minlength="10"
+              />
+              <div class="input-group-append">
+                <button class="btn btn-primary" id="sitePreferencesManualAdd" type="button">
```
Comment (varjolintu) [inline keepassxc-browser/options/options.html]: Warnings are not displayed when the value is invalid.

File: keepassxc-browser/options/options.html (hunk 8 above)
Comment (varjolintu) [reply keepassxc-browser/options/options.html]: Also, why the buttons are at the bottom, not after the input field? Then for example "Default group for saving new passwords" has the save button next to the input.

File: package.json (hunk 5 above)
Comment (thetric) [reply package.json]: Decreased minimum supported Firefox version to 52

File: keepassxc-browser/options/options.html (hunk 8 above)
Comment (thetric) [reply keepassxc-browser/options/options.html]: The buttons are now placed directly behind the inputs.
The "Blink Time" and "Redirect Offset" options look a bit weird with the `ms` text addon. Should it be removed?

Screenshot_2019-06-02_18-53-17

**Comment by thetric** [comment]
Remaining bugs:

* [ ] no form validation (blink timeout, site prefs > add url)
* [ ] site prefs/table: cannot change "Username-only Detection" on freshly added entry, although the changes are reflected to the model after a reload

File: keepassxc-browser/options/options.html (hunk 8 above)
Comment (varjolintu) [reply keepassxc-browser/options/options.html]: The "ms" is already explained in the help text.

**Comment by varjolintu** [comment]
We had a discussion and we would like to have this PR in the same release with https://github.com/keepassxreboot/keepassxc-browser/pull/351. The only changes concerning this PR are in the options page where the timeout settings have been removed totally. You should do the same. I'll modify the content script buttons to match the themes in this PR, because we cannot use Bootstrap there.

**Comment by thetric** [comment]
The timeout related settings UI + code is removed. The only remaining bug is the appearance of the checkbox in the "Site preferences" tab

**Comment by thetric** [comment]
I'm afraid I have no clue why the checkboxes don't toggle correctly. In [my version](https://github.com/thetric/keepassxc-browser/blob/bootstrap-4-upgrade/keepassxc-browser/options/options.js#L314) in the `change` handler of the checkbox the `url` is always `undefined` because `$(this)` refers to the *original* `tr.clone` but not to the *appended row*.

Might somebody point out the failure?

**Comment by varjolintu** [comment]
@thetric I'll look into this tomorrow and see if I can find the reason.

**Comment by varjolintu** [comment]
#772 replaces this. Closing the PR.
